### PR TITLE
Pattern reconciliation refactor

### PR DESCRIPTION
**Pull request by br648**
### Checklist

- [x] Appropriate branch selected _(all PRs must first be merged to `dev` before they can be merged to `master`)_
- [x] Any modified or new methods or classes have helpful JavaDoc and code is thoroughly commented
- [x] The description lists all applicable issues this PR seeks to resolve
- [x] The description lists any configuration setting(s) that differ from the default settings
- [x] All tests and CI builds passing

### Description

This PR addresses the issue of pattern stops and pattern locations being processed in series as part of the parent/child update in JDBCTableWriter. Pattern reconciliation (as well as pattern frequencies update) require all pattern stops and pattern locations to correctly update the stop times table.

JDBCTableWriter has been updated to collate pattern stops/locations and then triggers pattern reconciliation after the parent/child updates have been carried out. If pattern frequencies require updating this is triggered after pattern reconciliation so the cummulative travel times are correctly applied according to the updated pattern sequence.

This also addresses this bug: https://github.com/conveyal/gtfs-lib/pull/335#issuecomment-1079137841. See changes in src/main/java/com/conveyal/gtfs/model/StopTime.java.

### CODE REVIEW

File: src/main/java/com/conveyal/gtfs/loader/PatternReconciliation.java [hunk 1]
```diff
@@ -22,37 +23,87 @@ public class PatternReconciliation {
...
+        genericStops.sort(Comparator.comparingInt(genericStop -> genericStop.stopTime.stop_sequence));
+        return genericStops;
+    }
+
+    /**
+     * Define the type of update to be applied to stop times. This is done by comparing the size of the original generic
+     * stops against the size of the new generic stops.
+     */
+    private static ReconcileType getReconcileType(List<String> originalGenericStopIds, List<GenericStop> genericStops) {
+        int sizeDiff = originalGenericStopIds.size() - genericStops.size();
```
Comment (binh-dam-ibigroup) [inline src/main/java/com/conveyal/gtfs/loader/PatternReconciliation.java]: I am more used to having the difference is computed as `after - before`. That way, additions are positive numbers and that makes the if checks below more natural.

File: src/main/java/com/conveyal/gtfs/loader/PatternReconciliation.java [hunk 2]
```diff
@@ -22,37 +23,87 @@ public class PatternReconciliation {
...
+        if (sizeDiff == -1) {
+            return ReconcileType.ADD_ONE;
+        } else if (sizeDiff == 1) {
+            return ReconcileType.DELETE;
+        } else if (sizeDiff == 0) {
+            if (getFirstTripPatternDifference(originalGenericStopIds, genericStops) == -1) {
+                return ReconcileType.NONE;
+            }
+            return ReconcileType.TRANSPOSE;
```
Comment (binh-dam-ibigroup) [inline src/main/java/com/conveyal/gtfs/loader/PatternReconciliation.java]: A single `return` with a ternary expression could work here.

File: src/main/java/com/conveyal/gtfs/loader/PatternReconciliation.java [hunk 3]
```diff
@@ -65,35 +116,34 @@ public static void reconcilePatternLocations(
     private static void reconcilePattern(
         String tablePrefix,
         String patternId,
-        List<GenericStop> genericStops,
+        List<String> originalGenericStopIds,
+        List<GenericStop> newGenericStops,
         Connection connection,
-        PatternType patternType
+        ReconcileType reconcileType,
+        List<String> tripsForPattern
```
Comment (binh-dam-ibigroup) [inline src/main/java/com/conveyal/gtfs/loader/PatternReconciliation.java]: This is becoming kinda heavy on parameters. Maybe make `connection`, `tablePrefix`, `patternId`, `tripsForPattern` and other data that doesn't change during reconciliation be instance fields?

File: src/main/java/com/conveyal/gtfs/loader/PatternReconciliation.java [hunk 4]
```diff
@@ -173,7 +223,7 @@ private static void deleteOneStopFromAPattern(
     }
 
     /**
-     * Transpose two generic stops within a pattern.
+     * Move one generic stops within a pattern.
```
Comment (binh-dam-ibigroup) [inline src/main/java/com/conveyal/gtfs/loader/PatternReconciliation.java]: Stops singular.

File: src/main/java/com/conveyal/gtfs/util/json/JsonManager.java [hunk 5]
```diff
@@ -100,15 +92,31 @@ public String write (Map<String, T> map) throws JsonProcessingException {
...
+
+        List<T> list = new ArrayList<>();
+        for (JsonNode node : subEntities) {
+            ObjectNode objectNode = (ObjectNode) node;
+            if (!objectNode.get("id").isNumber()) {
+                // Set ID to zero. ID is ignored entirely here.
+                objectNode.put("id", 0);
+            }
+            // Accumulate new objects from JSON.
+            list.add(mapper.readValue(objectNode.toString(), target));
```
Comment (binh-dam-ibigroup) [inline src/main/java/com/conveyal/gtfs/util/json/JsonManager.java]: Would `mapper.treeToValue(objectNode, target)` be shorter?

File: src/main/java/com/conveyal/gtfs/loader/JdbcTableWriter.java [hunk 6]
```diff
@@ -272,6 +282,29 @@ public String update(Integer id, String json, boolean autoCommit) throws SQLExce
         }
     }
 
+    /**
+     * If an entity references a pattern (e.g., pattern stop or trip), determine whether the pattern uses
+     * frequencies because this impacts update behaviors, for example whether stop times are kept in
+     * sync with default travel times or whether frequencies are allowed to be nested with a JSON trip.
+     */
+    private boolean referencedPatternUsesFrequencies(Connection connection, ObjectNode jsonObject) throws SQLException {
+        boolean referencedPatternUsesFrequencies = false;
```
Comment (binh-dam-ibigroup) [inline src/main/java/com/conveyal/gtfs/loader/JdbcTableWriter.java]: Delete this var, place a `return` inside `while`, and `return false`  at the bottom of the method.

File: src/main/java/com/conveyal/gtfs/loader/JdbcTableWriter.java [hunk 7]
```diff
@@ -605,38 +638,11 @@ private String updateChildTable(
         if (!referencedPatternUsesFrequencies && subTable.name.equals(Table.FREQUENCIES.name) && subEntities.size() > 0) {
             // Do not permit the illegal state where frequency entries are being added/modified for a timetable pattern.
             throw new IllegalStateException("Cannot create or update frequency entries for a timetable-based pattern.");
+
```
Comment (binh-dam-ibigroup) [inline src/main/java/com/conveyal/gtfs/loader/JdbcTableWriter.java]: Remove extra blank line

File: src/main/java/com/conveyal/gtfs/loader/JdbcTableWriter.java [hunk 8]
```diff
@@ -605,38 +638,11 @@ private String updateChildTable(
...
-                if (!objectNode.get("id").isNumber()) {
-                    // Set ID to zero. ID is ignored entirely here. When the pattern stops are stored in the database,
-                    // the ID values are determined by auto-incrementation.
-                    objectNode.put("id", 0);
-                }
-                // Accumulate new pattern stop objects from JSON.
-                newPatternLocations.add(mapper.readValue(objectNode.toString(), PatternLocation.class));
-            }
-            PatternReconciliation.reconcilePatternLocations(tablePrefix, keyValue, newPatternLocations, connection);
+        boolean patternTables = Table.PATTERN_STOP.name.equals(subTable.name) || Table.PATTERN_LOCATION.name.equals(subTable.name);
```
Comment (binh-dam-ibigroup) [inline src/main/java/com/conveyal/gtfs/loader/JdbcTableWriter.java]: Rename to `isPatternTable`.

File: src/main/java/com/conveyal/gtfs/loader/JdbcTableWriter.java [hunk 9]
```diff
@@ -825,6 +799,117 @@ private String updateChildTable(
...
+        }
+    }
+
+    /**
+     * Pattern reconciliation requires all new pattern stops and pattern locations as well as the original values of both
+     * to correctly update stop times. Because these entities are processed in series, this method is used to accumulate
+     * all required values when available. The values are then used by {@link PatternReconciliation#reconcile} _after_
+     * all child entities have been processed.
+     */
+    private void stagePatternReconciliation(
```
Comment (binh-dam-ibigroup) [inline src/main/java/com/conveyal/gtfs/loader/JdbcTableWriter.java]: This could be a candidate to move into `PatternReconciliation` if you choose to add instance fields to it.

File: src/main/java/com/conveyal/gtfs/loader/JdbcTableWriter.java [hunk 10]
```diff
@@ -825,6 +799,117 @@ private String updateChildTable(
...
+        patternIdForReconciliation = keyValue;
+        if (originalGenericStopIds.isEmpty()) {
+            // Retrieve all generic stop ids before they are updated.
+            String sql = String.format("select location_id, stop_sequence from %s.pattern_locations pl " +
+                "where pl.pattern_id = ? " +
+                "union " +
+                "select stop_id, stop_sequence from %s.pattern_stops ps " +
+                "where ps.pattern_id = ? " +
+                "order by stop_sequence", tablePrefix, tablePrefix);
+            PreparedStatement getGenericStopIdsStatement = connection.prepareStatement(sql);
```
Comment (binh-dam-ibigroup) [inline src/main/java/com/conveyal/gtfs/loader/JdbcTableWriter.java]: It may be useful to make this prepared statement a field if it is called multiple times.

File: src/test/java/com/conveyal/gtfs/loader/JDBCTableWriterTest.java [hunk 11]
```diff
@@ -69,6 +84,7 @@ public static void setUpClass() throws SQLException, IOException, InvalidNamespa
         testDBName = TestUtils.generateNewDB();
         String dbConnectionUrl = String.format("jdbc:postgresql://localhost/%s", testDBName);
         testDataSource = TestUtils.createTestDataSource(dbConnectionUrl);
+        connection = testDataSource.getConnection();
```
Comment (binh-dam-ibigroup) [inline src/test/java/com/conveyal/gtfs/loader/JDBCTableWriterTest.java:105]: Does this need to be closed eventually?

File: src/test/java/com/conveyal/gtfs/loader/JDBCTableWriterTest.java [hunk 12]
```diff
@@ -69,6 +84,7 @@ public static void setUpClass() throws SQLException, IOException, InvalidNamespa
...
         LOG.info("creating feeds table because it isn't automatically generated unless you import a feed");
         Connection connection = testDataSource.getConnection();
```
Comment (binh-dam-ibigroup) [inline src/test/java/com/conveyal/gtfs/loader/JDBCTableWriterTest.java]: What is the difference between the two `connection` objects?

File: src/test/java/com/conveyal/gtfs/loader/JDBCTableWriterTest.java [hunk 13]
```diff
@@ -92,6 +108,45 @@ public static void setUpClass() throws SQLException, IOException, InvalidNamespa
...
+        pattern = createRouteAndPattern(
+            newUUID(),
+            patternId,
+            "pattern name",
+            null,
+            new ShapePointDTO[]{},
+            new PatternLocationDTO[]{
+                new PatternLocationDTO(patternId, locationOne.location_id, 0)
+            },
+            new PatternStopDTO[]{
```
Comment (binh-dam-ibigroup) [inline src/test/java/com/conveyal/gtfs/loader/JDBCTableWriterTest.java]: Insert space after `]` (multiple instances).

File: src/test/java/com/conveyal/gtfs/loader/JDBCTableWriterTest.java [hunk 14]
```diff
@@ -1024,340 +1079,203 @@ public void canUpdateServiceId() throws InvalidNamespaceException, IOException,
...
+    /**
+     * Various test cases for pattern reconciliation. These tests are applied to a single pattern
+     * see {@link JDBCTableWriterTest#patternReconciliationSetUp()} in the order defined.
+     */
+    private static Stream<PatternArguments> createPatternTests() {
+        return Stream.of(
+            // Add a new stop to the end.
+            new PatternArguments(
+                new PatternLocationDTO[]{
+                    new PatternLocationDTO(pattern.pattern_id, locationOne.location_id, 0, 10, 10)
```
Comment (binh-dam-ibigroup) [inline src/test/java/com/conveyal/gtfs/loader/JDBCTableWriterTest.java]: Extract var for `pattern.pattern_id`

File: src/test/java/com/conveyal/gtfs/loader/JDBCTableWriterTest.java [hunk 15]
```diff
@@ -1024,340 +1079,203 @@ public void canUpdateServiceId() throws InvalidNamespaceException, IOException,
...
+                new PatternStopDTO[]{
+                    new PatternStopDTO(pattern.pattern_id, stopOne.stop_id, 1, 10, 1),
+                    new PatternStopDTO(pattern.pattern_id, stopTwo.stop_id, 2, 10, 1)
+                },
+                ImmutableMap.of(
+                    locationOne.location_id, PatternReconciliation.PatternType.LOCATION,
+                    stopOne.stop_id, PatternReconciliation.PatternType.STOP,
+                    stopTwo.stop_id, PatternReconciliation.PatternType.STOP
+                ),
+                10,10,10,1
```
Comment (binh-dam-ibigroup) [inline src/test/java/com/conveyal/gtfs/loader/JDBCTableWriterTest.java]: Insert space after comma (multiple instances)

File: src/test/java/com/conveyal/gtfs/loader/JDBCTableWriterTest.java [hunk 16]
```diff
@@ -1024,340 +1079,203 @@ public void canUpdateServiceId() throws InvalidNamespaceException, IOException,
...
+        String sql = String.format(
+            "select * from %s.%s where trip_id='%s' and stop_id='%s' and stop_sequence=%s",
+            testNamespace,
+            Table.STOP_TIMES.name,
+            tripId,
+            stopId,
+            stopSequence
+        );
+        LOG.info(sql);
+        ResultSet stopTimesResultSet = connection.prepareStatement(sql).executeQuery();
```
Comment (binh-dam-ibigroup) [inline src/test/java/com/conveyal/gtfs/loader/JDBCTableWriterTest.java]: A regular statement can be used instead of a prepared statement.

**Review by binh-dam-ibigroup — COMMENTED** [review]
See my comments for now.

File: src/main/java/com/conveyal/gtfs/loader/PatternReconciliation.java (hunk 3 above)
Comment (br648) [reply src/main/java/com/conveyal/gtfs/loader/PatternReconciliation.java]: That has made the class a lot cleaner. Good spot!

File: src/test/java/com/conveyal/gtfs/loader/JDBCTableWriterTest.java (hunk 11 above)
Comment (br648) [reply src/test/java/com/conveyal/gtfs/loader/JDBCTableWriterTest.java:105]: Added `connection.close()` to tear down.

File: src/test/java/com/conveyal/gtfs/loader/JDBCTableWriterTest.java (hunk 12 above)
Comment (br648) [reply src/test/java/com/conveyal/gtfs/loader/JDBCTableWriterTest.java]: None. Have reduced to using one.

File: src/main/java/com/conveyal/gtfs/loader/JdbcTableWriter.java [hunk 17]
```diff
@@ -695,46 +681,15 @@ private String updateChildTable(
...
-            // Update linked stop times fields for each updated pattern stop (e.g., timepoint, pickup/drop off type).
-            if (Table.PATTERN_STOP.name.equals(subTable.name) || Table.PATTERN_LOCATION.name.equals(subTable.name)) {
-                if (referencedPatternUsesFrequencies) {
-                    // Update stop times linked to pattern stop if the pattern uses frequencies and accumulate time.
-                    // Default travel and dwell time behave as "linked fields" for associated stop times. In other
-                    // words, frequency trips in the editor must match the pattern stop travel times.
-                    int travelTimeForPatternHalts = Table.PATTERN_STOP.name.equals(subTable.name) ?
-                            updateStopTimesForPatternStop(subEntity, cumulativeTravelTime) :
-                            updateStopTimesForPatternLocation(subEntity, cumulativeTravelTime);
-                    cumulativeTravelTime += travelTimeForPatternHalts;
```
Comment (binh-dam-ibigroup) [inline src/main/java/com/conveyal/gtfs/loader/JdbcTableWriter.java:736]: Delete local var `cumulativeTravelTime` that's no longer used.

File: src/main/java/com/conveyal/gtfs/loader/JdbcTableWriter.java [hunk 18]
```diff
@@ -272,6 +264,28 @@ public String update(Integer id, String json, boolean autoCommit) throws SQLExce
...
+        if (jsonObject.has("pattern_id") && !jsonObject.get("pattern_id").isNull()) {
+            PreparedStatement statement = connection.prepareStatement(String.format(
+                "select use_frequency from %s.%s where pattern_id = ?",
+                tablePrefix,
+                Table.PATTERNS.name
+            ));
+            statement.setString(1, jsonObject.get("pattern_id").asText());
+            LOG.info(statement.toString());
+            ResultSet selectResults = statement.executeQuery();
+            while (selectResults.next()) {
```
Comment (binh-dam-ibigroup) [inline src/main/java/com/conveyal/gtfs/loader/JdbcTableWriter.java]: Convert to `if` (suggested by sonar-lint plugin).

File: src/main/java/com/conveyal/gtfs/loader/JdbcTableWriter.java [hunk 19]
```diff
@@ -272,6 +264,28 @@ public String update(Integer id, String json, boolean autoCommit) throws SQLExce
...
     }
 
+    /**
+     * If an entity references a pattern (e.g., pattern stop or trip), determine whether the pattern uses
+     * frequencies because this impacts update behaviors, for example whether stop times are kept in
+     * sync with default travel times or whether frequencies are allowed to be nested with a JSON trip.
+     */
+    private boolean referencedPatternUsesFrequencies(ObjectNode jsonObject) throws SQLException {
+        if (jsonObject.has("pattern_id") && !jsonObject.get("pattern_id").isNull()) {
+            PreparedStatement statement = connection.prepareStatement(String.format(
```
Comment (binh-dam-ibigroup) [inline src/main/java/com/conveyal/gtfs/loader/JdbcTableWriter.java]: Use try-with-resource on statement so it is closed.

File: src/main/java/com/conveyal/gtfs/loader/JdbcTableWriter.java [hunk 20]
```diff
@@ -171,9 +163,8 @@ public String update(Integer id, String json, boolean autoCommit) throws SQLExce
             if (specTable.name.equals("patterns")) {
                 referencingTables.add(Table.SHAPES);
             }
-
-            boolean referencedPatternUsesFrequencies = referencedPatternUsesFrequencies(connection, jsonObject);
-
+            PatternReconciliation.initialize(connection, tablePrefix);
```
Comment (binh-dam-ibigroup) [inline src/main/java/com/conveyal/gtfs/loader/JdbcTableWriter.java]: Oh, I thought you would write something like `PatternReconciliation reconciliation = new PatternReconciliation(...)`!

File: src/main/java/com/conveyal/gtfs/loader/PatternReconciliation.java [hunk 21]
```diff
@@ -37,46 +51,91 @@ public enum PatternType {
         LOCATION, STOP
     }
 
+    /**
+     * Set required parameters and clear any previously held data to prevent it impacting current processing.
+     */
+    public static void initialize(Connection connection, String tablePrefix) throws SQLException {
```
Comment (binh-dam-ibigroup) [inline src/main/java/com/conveyal/gtfs/loader/PatternReconciliation.java]: I would have written an instance constructor instead (with relevant fields and methods non-static)...

File: src/test/java/com/conveyal/gtfs/loader/JDBCTableWriterTest.java [hunk 22]
```diff
@@ -1257,7 +1258,7 @@ private void verifyStopTime(
             stopSequence
         );
         LOG.info(sql);
-        ResultSet stopTimesResultSet = connection.prepareStatement(sql).executeQuery();
+        ResultSet stopTimesResultSet = connection.createStatement().executeQuery(sql);
```
Comment (binh-dam-ibigroup) [inline src/test/java/com/conveyal/gtfs/loader/JDBCTableWriterTest.java]: Statement should be closed (use try-with-resources).

**Review by binh-dam-ibigroup — CHANGES_REQUESTED** [review]
Yep, much better, although I would convert most of the static members of `PatternReconciliation` into non-static.

File: src/main/java/com/conveyal/gtfs/loader/JdbcTableWriter.java (hunk 20 above)
Comment (br648) [reply src/main/java/com/conveyal/gtfs/loader/JdbcTableWriter.java]: er yeah, that does make more sense! Have addressed.

File: src/main/java/com/conveyal/gtfs/loader/JdbcTableWriter.java [hunk 23]
```diff
@@ -622,7 +630,7 @@ private String updateChildTable(
         }
         boolean isPatternTable = Table.PATTERN_STOP.name.equals(subTable.name) || Table.PATTERN_LOCATION.name.equals(subTable.name);
         if (isPatternTable) {
-            PatternReconciliation.stagePatternReconciliation(mapper, subTable, subEntities, keyValue);
+            reconciliation.stagePatternReconciliation(mapper, subTable, subEntities, keyValue);
```
Comment (binh-dam-ibigroup) [inline src/main/java/com/conveyal/gtfs/loader/JdbcTableWriter.java]: The method name can be shortened to `stage`.

File: src/main/java/com/conveyal/gtfs/loader/PatternReconciliation.java [hunk 24]
```diff
@@ -146,7 +166,7 @@ public static List<GenericStop> getGenericStops() {
      * Define the type of update to be applied to stop times. This is done by comparing the size of the original generic
      * stops against the size of the new generic stops.
      */
-    private static ReconcileType getReconcileType(List<String> originalGenericStopIds, List<GenericStop> genericStops) {
+    private ReconcileType getReconcileType(List<String> originalGenericStopIds, List<GenericStop> genericStops) {
```
Comment (binh-dam-ibigroup) [inline src/main/java/com/conveyal/gtfs/loader/PatternReconciliation.java]: Maybe rename to `getReconciliationOperation` or `getOperation` and rename the type to `ReconciliationOperation`.

File: src/test/java/com/conveyal/gtfs/loader/JDBCTableWriterTest.java [hunk 25]
```diff
@@ -1088,10 +1088,10 @@ private static Stream<PatternArguments> createPatternTests() {
         return Stream.of(
             // Add a new stop to the end.
             new PatternArguments(
-                new PatternLocationDTO[] {
+                new PatternLocationDTO[]{
```
Comment (binh-dam-ibigroup) [inline src/test/java/com/conveyal/gtfs/loader/JDBCTableWriterTest.java]: Revert these changes.

**Review by binh-dam-ibigroup — APPROVED** [review]
Thanks for the changes! A few minor tweaks, but otherwise looks good.

File: src/test/java/com/conveyal/gtfs/loader/JDBCTableWriterTest.java (hunk 25 above)
Comment (br648) [reply src/test/java/com/conveyal/gtfs/loader/JDBCTableWriterTest.java]: Updated my Intellij code style to match this.
